rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,132 @@
+# Integrate `cosign` with `kpack`
+
+## Problem
+
+The signing solution currently supported by `kpack` is
+[Notary v1](https://github.com/theupdateframework/notary). However, Notary
+presents itself with several different challenges, including:
+- It is unable to sign container image digests. The user must always specify a
+tag to be signed.
+- It is unable to verify container image signature by digests.
+- Signature relocation is not possible with this tool.
+- It is a project that currently has low activity and community involvement,
+  since the community is currently focused on the development of Notary v2.
+
+[`cosign`](https://github.com/sigstore/cosign) presents itself as a solution to
+some of those aforementioned challenges. With `cosign`, the user is able to:
+- Relocate container images across registries along with its signatures.
+  - Currently, there is no built-in command for relocation on `cosign`, but
+    copying container images and signatures across registries using other
+    container CLI tools such as `crane` and then verifying with `cosign` is
+    possible.
+  - There is also a proposal for
+    [implementing a `cosign cp`](https://github.com/sigstore/cosign/issues/303)
+    command that would affect this copy as a single operation.
+- Sign container images by tag or by digest.
+- Verify container image signatures by specifying either a tag or a digest.
+- Verify container image signatures after relocation.
+
+Moreover, `cosign` is a project that has had a lot of activity and community
+involvement recently.
+
+## Outcome
+
+`kpack` integrates with `cosign` to sign images it builds and push the
+signatures to a registry so that users can ensure the chain of custody of a
+generated artifact.
+
+This proposal aims to cover specifically the flow of signing an image produced
+by `kpack Image` resource without verification of base or builder images pulled
+in the process.
+
+## Actions to take
+
+### Enable image signing using `cosign`
+
+- Create a new configuration in the `Image` resource that allow users to
```

Comment:
Should we add an option to set the cosign registry or state that it will be supported?

It is currently controlled via `COSIGN_REPOSITORY`

Reply:
Makes sense to me. Does a single repository URL configuration address this concern, or should we add it as a part of the secret structure?

Reply:
@samj1912 how do you envision using `COSIGN_REPOSITORY`?

Reply:
@ncdc currently my usecase is more around the flexibility of having application tags being in one repository and the cosign tag in another so that it doesn't pollute the original repository with cosign tags and also so that I can control/manage the lifecycle of the repository with cosign signatures  differently than the application one. This also allows me to store signatures of images in an "operator" repository whereas the application images could be saved in a "user" repository.

Reply:
> Makes sense to me. Does a single repository URL configuration address this concern, or should we add it as a part of the secret structure?

Technically cosign allows each signature to be stored in a separate repository. It might be most flexible if we keep it along with the secret structure. But upto you and the maintainers.

Reply:
@matthewmcnew @tylerphelan would it make sense to combine this URL property with the annotations in a `ConfigMap`?

Reply:
@ncdc do you think it make more sense to move forward using `ConfigMap`s for the static annotations, taking into account @rszumlakowski's scenario, so we can combine static annotations and `COSIGN_REPOSITORY` URL in the same place?

Reply:
I think there is a valid use case for supporting `spec.cosign.signatureRepository` (`COSIGN_REPOSITORY`).

I also think that because a cosign signature supports annotations, it could make sense to have `spec.cosign.annotations`. (Unless there are good reasons not to support these per-`Image`?)

For static annotations that apply to multiple `Image`s, we can either try to design for that now, or we could defer while we continue to gather more use cases. My preference would be to defer.